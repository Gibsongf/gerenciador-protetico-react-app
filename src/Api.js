export async function apiLogin(loginData) {
    const url =
        "http://https://gerenciador-protetico.adaptable.app/users/login";
    // console.log(loginData);
    try {
        const response = await fetch(url, {
            method: "post",
            headers: {
                "Content-Type": "application/json",
            },
            body: JSON.stringify(loginData),
        });

        if (response.status === 200) {
            const data = await response.json();
            // console.log("Login successfully");
            localStorage.setItem("token", data.token);
            return true;
        } else {
            throw new Error();
        }
    } catch (error) {
        throw Error(error);
    }
}
export async function apiRegister(loginData) {
    const url =
        "http://https://gerenciador-protetico.adaptable.app/users/register";
    // console.log(loginData);
    try {
        const response = await fetch(url, {
            method: "post",
            headers: {
                "Content-Type": "application/json",
            },
            body: JSON.stringify(loginData),
        });

        if (response.status === 200) {
            const data = await response.json();
            // console.log("Login successfully");
            localStorage.setItem("token", data.token);
            return true;
        } else {
            throw new Error();
        }
    } catch (error) {
        throw Error(error);
    }
}
async function setupFetch(url, reqMethod = "get", body) {
    if (!localStorage["token"]) {
        await apiLogin();
        // localStorage.setItem("token", data.token);
    }

    const reqConfig = {
        method: reqMethod,
        headers: {
            "Content-Type": "application/json",
            Authorization: "Bearer " + localStorage["token"],
        },
    };
    if (body) {
        reqConfig.body = JSON.stringify(body);
    }
    try {
        const response = await fetch(url, reqConfig);
        if (response.status === 200 || response.status === 400) {
            const data = await response.json();
            return data;
        }

        if (response.status === 401) {
            localStorage.removeItem("token");
            return "Old token";
        }
    } catch (err) {
        throw Error(err);
    }
}
export async function APItodos(category) {
    const url = `http://https://gerenciador-protetico.adaptable.app/api/${category}/todos`;
    const data = await setupFetch(url, "get");
    // console.log(data);
    return data;
}
export async function APIDetails(category, id) {
    const url = `http://https://gerenciador-protetico.adaptable.app/api/${category}/${id}`;
    const data = await setupFetch(url, "get");
    return data;
}
export async function APIGetServiceBy(id, by) {
    const url = `http://https://gerenciador-protetico.adaptable.app/api/servico/todos/${by}/${id}`;
    const data = await setupFetch(url, "get");
    return data;
}

export async function APIPostNewData(formData) {
    const url = `http://https://gerenciador-protetico.adaptable.app/api/${formData["category"]}/novo`;
    // console.log(formData);
    const data = await setupFetch(url, "post", formData);
    // console.log(data);
    return data;
}
export async function APIPutData(formData, id) {
    const url = `http://https://gerenciador-protetico.adaptable.app/api/${formData["category"]}/${id}/edit`;
    // console.log(formData);
    const data = await setupFetch(url, "put", formData);
    // console.log(data);

    return data;
}
export async function ApiOneExcel(id) {
    const url = `http://https://gerenciador-protetico.adaptable.app/api/export/${id}`;
    const reqConfig = {
        method: "Get",
        headers: {
            Authorization: "Bearer " + localStorage["token"],
        },
    };

    try {
        const response = await fetch(url, reqConfig);
        if (response.status === 200) {
            const blob = await response.blob();
            // console.log(data);
            return blob;
        }
    } catch (err) {
        throw Error(err);
    }
}

export async function ApiMonthExcel(id, date) {
    const url = `http://https://gerenciador-protetico.adaptable.app/api/export/${id}/${date}`;
    const reqConfig = {
        method: "Get",
        headers: {
            Authorization: "Bearer " + localStorage["token"],
        },
    };

    try {
        const response = await fetch(url, reqConfig);
        if (response.status === 200) {
            const blob = await response.blob();
            // console.log(data);
            return blob;
        }
    } catch (err) {
        throw Error(err);
    }
}
