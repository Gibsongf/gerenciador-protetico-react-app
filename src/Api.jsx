export async function apiLogin(loginData) {
    const url = "http://localhost:3000/users/login";
    // console.log(loginData);
    try {
        const response = await fetch(url, {
            method: "post",
            headers: {
                "Content-Type": "application/json",
            },
            body: JSON.stringify(loginData),
        });

        if (response.status === 200) {
            const data = await response.json();
            console.log("Login successfully");
            localStorage.setItem("token", data.token);
            return true;
        } else {
            throw new Error();
        }
    } catch (error) {
        throw Error(error);
    }
}

async function setupFetch(url, reqMethod = "get", body) {
    if (!localStorage["token"]) {
        await apiLogin();
        // localStorage.setItem("token", data.token);
    }

    const reqConfig = {
        method: reqMethod,
        headers: {
            "Content-Type": "application/json",
            Authorization: "Bearer " + localStorage["token"],
        },
    };
    if (body) {
        reqConfig.body = JSON.stringify(body);
    }
    try {
        const response = await fetch(url, reqConfig);
        if (response.status === 200 || response.status === 400) {
            const data = await response.json();
            return data;
        }

        if (response.status === 401) {
            localStorage.removeItem("token");
            return "Old token";
        }
    } catch (err) {
        throw Error(err);
    }
}
export async function APItodos(category) {
    const url = `http://localhost:3000/api/${category}/todos`;
    const data = await setupFetch(url, "get");
    // console.log(data);
    return data;
}
export async function APIDetails(category, id) {
    const url = `http://localhost:3000/api/${category}/${id}`;
    const data = await setupFetch(url, "get");
    return data;
}
export async function APIGetServiceBy(id, by) {
    const url = `http://localhost:3000/api/servico/todos/${by}/${id}`;
    const data = await setupFetch(url, "get");
    return data;
}

export async function APIPostNewData(formData) {
    const url = `http://localhost:3000/api/${formData["category"]}/novo`;
    console.log(formData);
    const data = await setupFetch(url, "post", formData);
    // console.log(data);
    return data;
}
export async function APIPutData(formData, id) {
    const url = `http://localhost:3000/api/${formData["category"]}/${id}/edit`;
    // console.log(formData);
    const data = await setupFetch(url, "put", formData);
    // console.log(data);

    return data;
}
export async function ExcelLink(id) {
    const url = `http://localhost:3000/api/export/${id}`;
    const reqConfig = {
        method: "Get",
        headers: {
            Authorization: "Bearer " + localStorage["token"],
        },
    };

    try {
        const response = await fetch(url, reqConfig);
        if (response.status === 200) {
            const blob = await response.blob();
            // console.log(data);
            return blob;
        }
    } catch (err) {
        throw Error(err);
    }
}
