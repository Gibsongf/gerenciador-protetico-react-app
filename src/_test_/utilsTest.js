export const testService = {
    paciente: "Meire Macedo",
    dentista: "dentista-1",
    local: "local-1",
    produto: [
        {
            _id: "65dd5ff530d103b5e6f12059",
            nome: "Sem marca Granito Teclado",
            valor_normal: 132,
            valor_reduzido: 66,
            __v: 0,
        },
        {
            _id: "65dd5ff530d103b5e6f12053",
            nome: "Ergonômico Madeira Queijo",
            valor_normal: 200,
            valor_reduzido: 67,
            __v: 0,
        },
    ],
    statusEntrega: false,
    category: "servico",
    formType: "edit",
    dbId: "65dd5ff530d103b5e6f12067",
};
export const testDentist = {
    nome: "first dentist",
    sobrenome: "first sobrenome",
    telefone: "89224002",
    cpf: "10987654321",
    category: "dentista",
    formType: "edit",
    _id: "first-dentist-id",
    serviço: [],
};
export const testDentist2 = {
    nome: "second dentist",
    sobrenome: "second sobrenome",
    telefone: "40028922",
    cpf: "12345678910",
    category: "dentista",
    formType: "edit",
    _id: "second-dentist-id",
    serviço: [],
};
export const testLocal = [
    {
        _id: "local-0",
        nome: "Franco EIRELI",
        endereço: "333 Felícia Marginal",
        cep: "23147-272",
        telefone: "1174-0086",
        tabela: "Reduzido",
        dentistas: [testDentist],
        __v: 0,
    },
    {
        _id: "local-1",
        nome: "Moraes LTDA",
        endereço: "1047 Luiza Alameda",
        cep: "63006-953",
        telefone: "5541-7666",
        tabela: "Normal",
        dentistas: [testDentist2],
        __v: 0,
    },
];
