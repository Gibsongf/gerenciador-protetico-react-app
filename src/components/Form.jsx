import { useState } from "react";
import {
    CepInput,
    CpfInput,
    SelectInput,
    SimpleInput,
    TelefoneInput,
    TipoTabelaSelect,
} from "./Inputs";
import "../styles/Forms.css";
const UseForm = (formType, initialState) => {
    const [formData, setFormData] = useState(initialState);

    const handleChange = (e) => {
        // console.log(e.target);
        // console.log(formData);
        setFormData({ ...formData, [e.target.name]: e.target.value });
        // newContentValidator(e);
    };
    const handleSubmit = async (e) => {
        e.preventDefault();
        // const validData = newContentValidator(e.target.parentElement);

        // if (validData) {
        //     if (formType === "new") {
        //         // await newPost(formData);
        //         setFormData({ text: "", title: "", published: false });
        //     }
        //     if (formType === "edit") {
        //         setWasUpdated((e) => !e);
        //     }
        // }
    };
    return { formData, handleChange, handleSubmit };
};

export function FormLocal() {
    const nomeTag = { id: "nome", txt: "Nome do Local ou a Franquia:" };
    const initialState = {
        nome: "",
        endereço: "",
        cep: "",
        telefone: "",
        tabela: "",
    };
    const { formData, handleChange, handleSubmit } = UseForm(
        "new",
        initialState
    );

    return (
        <form action="">
            <legend>
                <h3>Registrar Novo Local</h3>
            </legend>
            <SimpleInput
                id={nomeTag.id}
                labelTxt={nomeTag.txt}
                value={formData.nome}
                onChange={handleChange}
            />
            <SimpleInput
                id={"endereço"}
                labelTxt={"Endereço:"}
                value={formData.endereço}
                onChange={handleChange}
            />
            <CepInput value={formData.cep} onChange={handleChange} />
            <TelefoneInput value={formData.telefone} onChange={handleChange} />
            <TipoTabelaSelect value={formData.tabela} onChange={handleChange} />
            <button type="submit">Registrar</button>
        </form>
    );
}
export function FormDentist() {
    const initialState = {
        nome: "",
        sobrenome: "",
        local: "",
        telefone: "",
        cpf: "",
    };

    const { formData, handleChange, handleSubmit } = UseForm(
        "new",
        initialState
    );

    return (
        <form action="">
            <legend>
                <h3>Registrar Novo Dentista</h3>
            </legend>
            <SimpleInput
                id={"nome"}
                labelTxt={"Nome:"}
                value={formData.nome}
                onChange={handleChange}
            />
            <SimpleInput
                id={"sobrenome"}
                labelTxt={"Sobrenome:"}
                value={formData.sobrenome}
                onChange={handleChange}
            />
            <TelefoneInput value={formData.telefone} onChange={handleChange} />
            <CpfInput value={formData.cpf} onChange={handleChange} />

            <SelectInput
                initialValue={formData.local}
                onChange={handleChange}
                category={"local"}
                labelTxt={"Local de Trabalho:"}
            />
            <button type="submit">Registrar</button>
        </form>
    );
}

export function FormProduct() {
    const initialState = {
        nome: "",
        valor_normal: "",
        valor_reduzido: "",
    };

    const { formData, handleChange, handleSubmit } = UseForm(
        "new",
        initialState
    );

    return (
        <form action="">
            <legend>
                <h3>Registrar Novo Produto</h3>
            </legend>
            <SimpleInput
                id={"nome"}
                labelTxt={"Nome:"}
                value={formData.nome}
                onChange={handleChange}
            />
            <SimpleInput
                id={"valor_normal"}
                labelTxt={"Valor Normal:"}
                value={formData.valor_normal}
                onChange={handleChange}
                type={"number"}
            />
            <SimpleInput
                id={"valor_reduzido"}
                labelTxt={"Valor Reduzido:"}
                value={formData.valor_reduzido}
                onChange={handleChange}
                type={"number"}
            />

            <button type="submit">Registrar</button>
        </form>
    );
}

export function FormService() {
    const initialState = {
        dentista: "",
        produto: "",
        paciente_nome: "",
    };

    const { formData, handleChange, handleSubmit } = UseForm(
        "new",
        initialState
    );

    return (
        <form action="">
            <legend>
                <h3>Registrar Novo Serviço</h3>
            </legend>
            <SimpleInput
                id={"paciente_nome"}
                labelTxt={"Nome do Paciente:"}
                value={formData.paciente_nome}
                onChange={handleChange}
            />
            {/* select Local then from this local we 
                show the dentist that work there as
                option for the next select input  
            */}
            <SelectInput
                initialValue={formData.dentista}
                onChange={handleChange}
                category={"local"}
                labelTxt={"Local:"}
            />

            <button type="submit">Registrar</button>
        </form>
    );
}
