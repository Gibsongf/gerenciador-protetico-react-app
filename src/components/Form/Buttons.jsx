import { useContext } from "react";
import PropTypes from "prop-types";
import { EditContext } from "../../pages/Details";

ButtonRegister.propTypes = {
    handleSubmit: PropTypes.func,
};
ButtonEdit.propTypes = {
    handleSubmit: PropTypes.func,
};
export function ButtonRegister({ handleSubmit }) {
    return (
        <button onClick={handleSubmit} type="submit">
            Registrar
        </button>
    );
}
export function ButtonEdit({ handleSubmit }) {
    const { setEdit } = useContext(EditContext);
    return (
        <>
            <button onClick={handleSubmit} type="submit">
                Confirm
            </button>

            <button type="button" onClick={() => setEdit((e) => !e)}>
                Cancel
            </button>
        </>
    );
}
