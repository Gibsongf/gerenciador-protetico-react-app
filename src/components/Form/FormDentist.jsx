import { useContext, useRef, useState } from "react";

import { CpfInput, SelectInput, SimpleInput, TelefoneInput } from "../Inputs";
import "../../styles/Forms.css";
import PropTypes from "prop-types";

import { useForm } from "./useForm";
import { AppContext } from "../../App";
import { ButtonEdit, ButtonRegister } from "./Buttons";
import { NewFormContext } from "../NewFormButton";
import { EditContext } from "../GenerateDetails";

export function FormDentist({ initialState, closeBtn }) {
    const ref = useRef();
    const { errorMsg } = useContext(AppContext);
    const { setClose, setTableUpdate } = useContext(NewFormContext);
    const { setEdit, setUpdate } = useContext(EditContext);
    let legendTxt = "Editar Detalhes do Dentista";
    if (!initialState) {
        legendTxt = "Registrar Novo Dentista";
        initialState = {
            nome: "",
            sobrenome: "",
            local: "",
            telefone: "",
            cpf: "",
            category: "dentista",
            formType: "new",
        };
    }

    const { formData, handleChange, handleSubmit } = useForm(
        initialState,
        ref.current
    );
    const handleSubmitResponse = async (e) => {
        e.preventDefault();

        const success = await handleSubmit(e);
        if (success) {
            if (initialState.formType === "new") {
                setClose((e) => !e);
                setTableUpdate((e) => !e);
            }
            if (initialState.formType === "edit") {
                setEdit((e) => !e);
                setUpdate((e) => !e);
            }
        }
    };
    return (
        <div className="form-container" id="pop-up">
            <form action="" ref={ref} id="pop-up-content">
                {closeBtn}
                <legend>
                    <h3>{legendTxt}</h3>
                </legend>
                <SimpleInput
                    id={"nome"}
                    labelTxt={"Nome"}
                    value={formData.nome}
                    onChange={handleChange}
                    msg={!errorMsg ? "" : errorMsg.nome}
                />
                <SimpleInput
                    id={"sobrenome"}
                    labelTxt={"Sobrenome"}
                    value={formData.sobrenome}
                    onChange={handleChange}
                />
                <TelefoneInput
                    value={formData.telefone}
                    onChange={handleChange}
                    msg={!errorMsg ? "" : errorMsg.telefone}
                />
                <CpfInput
                    value={formData.cpf}
                    onChange={handleChange}
                    msg={!errorMsg ? "" : errorMsg.cpf}
                />

                <SelectInput
                    initialValue={formData.local}
                    onChange={handleChange}
                    category={"local"}
                    labelTxt={"Local de Trabalho"}
                    msg={!errorMsg ? "" : errorMsg.local}
                />
                {initialState.formType === "edit" && (
                    <ButtonEdit handleSubmit={handleSubmitResponse} />
                )}
                {initialState.formType === "new" && (
                    <ButtonRegister handleSubmit={handleSubmitResponse} />
                )}
            </form>
        </div>
    );
}

FormDentist.propTypes = {
    initialState: PropTypes.object,
    closeBtn: PropTypes.any,
};
