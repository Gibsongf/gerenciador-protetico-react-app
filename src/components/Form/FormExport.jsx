import { createContext, useContext, useRef } from "react";

import {
    EntregaStatus,
    FilterDentistsByLocation,
    SearchProducts,
    SelectInput,
    SimpleInput,
} from "../Inputs";
import PropTypes from "prop-types";
import "../../styles/Forms.css";
import { useTodosApi } from "../ApiHooks";
import { AppContext } from "../../App";
import { useForm } from "./useForm";
import { ButtonConfirm, ButtonEdit, ButtonRegister } from "../Buttons";
import { NewFormContext } from "../NewFormButton";
import { EditContext } from "../GenerateDetails";
import { PopUpEditContext } from "../../pages/Todos";

FormExport.propTypes = {
    initialState: PropTypes.any,
    closeBtn: PropTypes.any,
};
export const RefContext = createContext({
    checkBoxRef: "",
});
export function FormExport({ initialState, closeBtn }) {
    const { setExport } = useContext(PopUpEditContext);

    // const products = useTodosApi("produto");
    // const { formData, setFormData, handleChange, handleSubmit } = useForm(
    //     initialState,
    //     ref.current
    // );

    // use the ref context to filter the selected checkbox and
    //format to array and save to the form data

    // console.log(initialState, formData);
    const beforeSendSubmit = async (e) => {
        e.preventDefault();
        //before submit we format the product

        // const success = await handleSubmit(e);
        // if (success) {
        //     if (initialState.formType === "new") {
        //         setClose((e) => !e);
        //         setTableUpdate((e) => !e);
        //     }
        //     if (initialState.formType === "edit") {
        //         setShowForm(() => false);
        //         setUpdate((e) => !e);
        //     }
        // }
    };

    return (
        <div className="form-container" id="pop-up">
            <form action="" id="pop-up-content">
                {closeBtn}
                <legend>
                    <h3>Configurar Exportação</h3>
                </legend>
                <label htmlFor="month">
                    / Mês:{" "}
                    <input
                        style={{ width: "160px" }}
                        // ref={ref}
                        // onChange={onChangeMonth}
                        lang="pt-BR"
                        type="month"
                        name="month"
                        id="month"
                    />
                </label>
                {/* <SimpleInput
                    id={"paciente"}
                    labelTxt={"Nome do Paciente:"}
                    value={formData.paciente}
                    onChange={handleChange}
                    msg={!errorMsg ? "" : errorMsg.paciente}
                /> */}

                {/* <SelectInput
                    initialValue={formData.local}
                    onChange={handleChange}
                    category={"local"}
                    labelTxt={"Local:"}
                    msg={!errorMsg ? "" : errorMsg.local}
                /> */}

                {/* <FilterDentistsByLocation
                    initialValue={formData.dentista}
                    onChange={handleChange}
                    dbId={formData.local}
                    msg={!errorMsg ? "" : errorMsg.dentista}
                /> */}

                <EntregaStatus
                    value={formData.statusEntrega}
                    onChange={handleChange}
                />
            </form>
        </div>
    );
}
