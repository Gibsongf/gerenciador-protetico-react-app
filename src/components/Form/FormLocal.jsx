import { useContext, useRef } from "react";
import {
    CepInput,
    SimpleInput,
    TelefoneInput,
    TipoTabelaSelect,
} from "../Inputs";
import "../../styles/Forms.css";
import { AppContext } from "../../App";
import { useForm } from "./useForm";
import PropTypes from "prop-types";
import { ButtonClose, ButtonEdit, ButtonRegister } from "../Buttons";
import { NewFormContext } from "../NewFormButton";

export function FormLocal({ initialState, setEdit, setUpdate }) {
    const ref = useRef();
    const { errorMsg } = useContext(AppContext);
    const nomeTag = { id: "nome", txt: "Nome do Local:" };
    const { setClose, setTableUpdate } = useContext(NewFormContext);

    let legendTxt = "Editar Detalhes do Local";
    if (!initialState) {
        legendTxt = "Registrar Novo Local";
        initialState = {
            nome: "",
            endereço: "",
            cep: "",
            telefone: "",
            tabela: "Normal",
            category: "local",
            formType: "new",
        };
    }
    const { formData, handleChange, handleSubmit } = useForm(
        initialState,
        ref.current
    );
    const handleSubmitResponse = async (e) => {
        const success = await handleSubmit(e);
        if (success) {
            if (initialState.formType === "new") {
                setClose((e) => !e);
                setTableUpdate((e) => !e);
            }
            if (initialState.formType === "edit") {
                setEdit((e) => !e);
                setUpdate((e) => !e);
            }
        }
    };

    return (
        <div className="form-container" id="pop-up">
            <form
                onSubmit={handleSubmitResponse}
                action=""
                ref={ref}
                id="pop-up-content"
            >
                <ButtonClose
                    setClose={
                        initialState.formType === "new" ? setClose : setEdit
                    }
                />
                <legend>
                    <h3>{legendTxt}</h3>
                </legend>

                <SimpleInput
                    id={nomeTag.id}
                    labelTxt={nomeTag.txt}
                    value={formData.nome}
                    onChange={handleChange}
                    msg={!errorMsg ? "" : errorMsg.nome}
                />
                <SimpleInput
                    id={"endereço"}
                    labelTxt={"Endereço:"}
                    value={formData.endereço}
                    onChange={handleChange}
                    msg={!errorMsg ? "" : errorMsg.endereço}
                />
                <CepInput
                    value={formData.cep}
                    onChange={handleChange}
                    msg={!errorMsg ? "" : errorMsg.cep}
                />
                <TelefoneInput
                    value={formData.telefone}
                    onChange={handleChange}
                    msg={!errorMsg ? "" : errorMsg.telefone}
                />
                <TipoTabelaSelect
                    value={formData.tabela}
                    onChange={handleChange}
                />
                {initialState.formType === "edit" && <ButtonEdit />}
                {initialState.formType === "new" && <ButtonRegister />}
            </form>
        </div>
    );
}

FormLocal.propTypes = {
    initialState: PropTypes.object,
    setEdit: PropTypes.any,
    setUpdate: PropTypes.any,
};
