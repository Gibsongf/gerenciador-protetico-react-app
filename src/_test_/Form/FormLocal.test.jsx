/* eslint-disable no-undef */
import { render, screen } from "@testing-library/react";
import { expect, vi } from "vitest";
import userEvent from "@testing-library/user-event";
import { FormLocal } from "../../components/Form/FormLocal";
import { APIPostNewData, APIPutData } from "../../Api";
import { mockServiceData } from "../utilsTest";

//need to get id of some local and mock useForm return a local?
describe("Form Local component", () => {
    //all useful elements of the form
    const formInfo = mockServiceData.local[0];
    const expectFormElements = (el) => {
        expect(el.local.value).toBe(formInfo.nome);
        expect(el.endereço.value).toBe(formInfo["endereço"]);
        expect(el.cep.value).toBe(formInfo.cep);
        expect(el.telefone.value).toBe(formInfo.telefone);
    };
    const getEl = () => {
        const local = screen.getByLabelText("Nome do Local");
        const endereço = screen.getByLabelText("Endereço");
        const cep = screen.getByLabelText("Cep");
        const telefone = screen.getByLabelText("Telefone");
        const tabela = screen.getByLabelText("Tipo de Tabela");
        const button = screen.getAllByRole("button")[1];
        return { local, endereço, cep, telefone, tabela, button };
    };

    const inputText = async (el, text, user) => {
        await user.dblClick(el);
        await user.keyboard(text);
    };
    it("renders correctly", () => {
        const { container } = render(<FormLocal />);
        const form = screen.getByRole("local-form");
        expect(container).toMatchSnapshot();
        expect(form).toBeInTheDocument();
    });
    it("register all input, submit form & call API", async () => {
        const user = userEvent.setup();
        render(<FormLocal />);
        const { local, endereço, cep, telefone, button } = getEl();
        const header = screen.getByRole("heading");
        const tabelaSelector = screen.getByLabelText("Tipo de Tabela");

        await inputText(local, formInfo.nome, user);
        await inputText(endereço, formInfo["endereço"], user);
        await inputText(cep, formInfo.cep, user);
        await inputText(telefone, formInfo.telefone, user);
        await user.selectOptions(tabelaSelector, ["Reduzido"]);
        await user.click(button);

        //header without initialState
        expect(header.textContent).toBe("Registrar Novo Local");
        //input with expected values
        expectFormElements({ local, endereço, cep, telefone });
        expect(tabelaSelector.value).toBe("Reduzido");
        expect(APIPostNewData).toBeCalled();
    });
    it("Form edit mode, save data at right inputs", async () => {
        const user = userEvent.setup();
        const setUpdate = vi.fn();
        const setEdit = vi.fn();

        render(
            <FormLocal
                initialState={formInfo}
                setEdit={setEdit}
                setUpdate={setUpdate}
            />
        );
        const { local, endereço, cep, telefone, button } = getEl();
        const header = screen.getByRole("heading");
        await user.click(button);

        //input with expected values
        expectFormElements({ local, endereço, cep, telefone });
        //heading with initialState
        expect(header.textContent).toBe("Editar Detalhes do Local");
        //call api update data
        expect(APIPutData).toBeCalled();
        //when success
        //close the form element
        expect(setEdit).toBeCalled();
        //will update current page info
        expect(setUpdate).toBeCalled();
    });
    it("submit empty form, won't call API", async () => {
        const user = userEvent.setup();
        render(<FormLocal />);
        const { button } = getEl();
        await user.click(button);

        expect(APIPostNewData).not.toBeCalled();
    });
});
