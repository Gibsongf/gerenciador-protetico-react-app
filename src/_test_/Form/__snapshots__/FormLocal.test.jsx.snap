// Vitest Snapshot v1, https://vitest.dev/guide/snapshot.html

exports[`Form Local component > renders correctly 1`] = `
<div>
  <div
    class="form-container"
    id="pop-up"
  >
    <form
      action=""
      id="pop-up-content"
      role="local-form"
    >
      <span
        class="form-close-btn"
      >
        <button
          class="close-btn"
          type="button"
        >
          X
        </button>
      </span>
      <legend>
        <h3>
          Registrar Novo Local
        </h3>
      </legend>
      <div
        class="nome"
      >
        <label
          for="nome"
        >
          Nome do Local
        </label>
        <div
          class="error-message"
        />
        <input
          id="nome"
          name="nome"
          required=""
          type="text"
          value=""
        />
      </div>
      <div
        class="endereço"
      >
        <label
          for="endereço"
        >
          Endereço
        </label>
        <div
          class="error-message"
        />
        <input
          id="endereço"
          name="endereço"
          required=""
          type="text"
          value=""
        />
      </div>
      <div
        class="Cep"
      >
        <label
          for="cep"
        >
          Cep
        </label>
        <div
          class="error-message"
        />
        <input
          id="cep"
          name="cep"
          placeholder="xxxxx-xxx"
          required=""
          type="text"
          value=""
        />
      </div>
      <div>
        <label
          for="telefone"
        >
          Telefone
        </label>
        <div
          class="error-message"
        />
        <input
          id="telefone"
          maxlength="9"
          name="telefone"
          placeholder="xxxxx-xxxx"
          required=""
          type="tel"
          value=""
        />
      </div>
      <div
        class="tabela-options"
      >
        <label
          for="tabela-options"
        >
          Tipo de Tabela
        </label>
        <select
          id="tabela-options"
          name="tabela"
        >
          <option
            value="Normal"
          >
            Normal
          </option>
          <option
            value="Reduzido"
          >
            Reduzido
          </option>
        </select>
      </div>
      <button
        class="register-form-btn"
        type="submit"
      >
        Registrar
      </button>
    </form>
  </div>
</div>
`;
